extensions/flight-sql/src/main/java/io/deephaven/server/flightsql/FlightSqlResolver.java

```diff
@@ -0,0 +1,1719 @@
+            if (!preparedStatements.remove(handleId, this)) {
+                return;
+            }
+            for (CommandPreparedStatementQueryImpl query : queries) {
+                query.releaseImpl(false);
+            }
+            queries.clear();
+        }
+    }
+
+    /**
+     * The Arrow "specification" for filter pattern leaves a lot to be desired. In totality:
+     *
+     * <pre>
+     * In the pattern string, two special characters can be used to denote matching rules:
+     *    - "%" means to match any substring with 0 or more characters.
+     *    - "_" means to match any one character.
+     * </pre>
+     *
+     * There does not seem to be any potential for escaping, which means that underscores can't explicitly be matched
+     * against, which is a common pattern used in Deephaven table names. As mentioned below, it also follows that an
+     * empty string should only explicitly match against an empty string.
+     */
+    private static Predicate<String> flightSqlFilterPredicate(String flightSqlPattern) {
+        // This is the technically correct, although likely represents a Flight SQL client mis-use, as the results will
+        // be empty (unless an empty db_schema_name is allowed).
+        //
+        // Unlike the "catalog" field in CommandGetDbSchemas (/ CommandGetTables) where an empty string means
+        // "retrieves those without a catalog", an empty filter pattern does not seem to be meant to match the
+        // respective field where the value is not present.
+        //
+        // The Arrow schema for CommandGetDbSchemas explicitly points out that the returned db_schema_name is not null,
+        // which implies that filter patterns are not meant to match against fields where the value is not present
+        // (null).
+        if (flightSqlPattern.isEmpty()) {
+            // If Deephaven supports catalog / db_schema_name in the future and db_schema_name can be empty, we'd need
+            // to match on that.
+            // return String::isEmpty;
+            return x -> false;
+        }
+        if ("%".equals(flightSqlPattern)) {
+            return x -> true;
+        }
+        if (flightSqlPattern.indexOf('%') == -1 && flightSqlPattern.indexOf('_') == -1) {
+            // If there are no special characters, search for an exact match; this case was explicitly seen via the
+            // Flight SQL JDBC driver.
+            return flightSqlPattern::equals;
+        }
+        final int L = flightSqlPattern.length();
+        final StringBuilder pattern = new StringBuilder();
+        final StringBuilder quoted = new StringBuilder();
+        final Runnable appendQuoted = () -> {
+            if (quoted.length() != 0) {
+                pattern.append(Pattern.quote(quoted.toString()));
+                quoted.setLength(0);
+            }
+        };
+        for (int i = 0; i < L; ++i) {
+            final char c = flightSqlPattern.charAt(i);
```

Comment:
At least right now, the table names are limited to the sorts of names we allow in the query scope.

Reply:
Bumping this - this may still be wrong. 

Both Python and Java permit multi-codeunit codepoints, such as `𞸷`, to be identifiers. This can't be evaluated in our console at this time, but if it can be bound to an identifier, you can reference it as a table name, and char count does not equal codepoint count:
```
jshell> "𞸷".length()
$17 ==> 2

jshell> "𞸷".codePoints().count()
$18 ==> 1
```

I can't find an example quickly where the second `char`/codeunit can be something we want to match and so must never split, but the fix is easy and obviously won't have this issue - either scan with a Pattern for those two strings (and thus let the Matcher tell us the substrings instead of manually scanning char by char with a lambda to append), or iterate along the `String.codePoints()` stream and do your own map/reduce like you're doing now.

Reply:
Posted in another comment as well, but the FlightSQL implementation itself iterates through charAt; not that that absolves us of doing the right thing....

https://github.com/apache/arrow/blob/apache-arrow-18.0.0/java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/ArrowDatabaseMetadata.java#L1253-L1277

Reply:
It's been a while since I've done codepoint stuff. I don't know if it's worthwhile spending time now getting it "right", given that 1) The FlightSQL implementation above has the same behavior, and 2) the "spec" itself seems to leave a lot to be desired (I wouldn't be surprised if they scrap it and come up with something better in the future).

I can file a DHC issue, or even Arrow issue?

Reply:
I don't think this is an arrow issue, it is nothing at all to do with the (poor) choice of wildcard tokens and lack of an escaping mechanism.

Instead it is more akin to "this is a stream of ints, but we're reading it byte-by-byte and if a byte matches, we doSomething()" and thus truncate an int unintentionally. Changing how you delimit the data (either walk codepoints themselves, or split the string into proper substrings) makes this go away.

Reply:
Latest push is iterating over the code points, and should correctly handle non BMP code points.